lib/common/abstract-descriptor.h

```diff
@@ -0,0 +1,269 @@
+
+  // RankedSizedArray<A>(n) are arrays with storage for at least n A elements.
+  // The implementation is free to allocate more memory space than needed.
+  // n must not be greater than maxRank;
+  template<typename A>
+  using RankedSizedArray = typename DESC::template RankedSizedArray<A>;
+
+  // SubscriptArray is an alias for RankedSizedArray<SubscriptValue>
+  using SubscriptArray = typename DESC::SubscriptArray;
+
+  // OwningPointer<A> acts like a A*. When receiving an OwningPointer<A>,
+  // one is responsible to destroy its content. It may be implemented as
+  // a simple A* or something more complex and safer.
+  template<typename A>
+  using OwningPointer = typename DESC::template OwningPointer<A>;
+
+  // TODO: Describe FortranType and Dimension interface or hide them
+  // FortranType holds type information
+  using FortranType = typename DESC::FortranType;
+  // Dimension holds dimension information.
+  using Dimension = typename DESC::Dimension;
+
+  // Constants to be defined by descriptor implementation
+  static constexpr Rank maxRank{DESC::maxRank};
+
+  // A descriptor implementation must provide the following functions:
+
+  Rank rank() const { return DESC::rank(); }
+
+  FortranType type() const { return DESC::type(); }
+
+  Dimension GetDimension(Rank i) const { return DESC::GetDimension(i); }
+
+  // Write lower bounds of the entity describe by the descriptor
+  // into subscripts.
+  void GetLowerBounds(SubscriptArray &subscripts) const {
+    DESC::GetLowerBounds(subscripts);
+  }
+
+  bool HasSameTypeAs(const DescriptorInterface<DESC> &other) const {
+    return DESC::HasSameTypeAs(static_cast<const DESC &>(other));
+  }
+
+  // When the decsriptor describes an array containing integer,
+  // GetSubscriptValueAt extract the element indexed by subscripts and convert
+  // it to a SubscriptValue.
+  SubscriptValue GetSubscriptValueAt(const SubscriptArray &subscripts) const {
+    return DESC::GetSubscriptValueAt(subscripts);
+  }
+
+  // Number of elements
+  SizeValue Elements() const { return DESC::Elements(); }
+
+  // Create a descriptor with the same type (including type parameters if
+  // applicable) as source and the given rank. The data and bounds are left
+  // unset.
+  static OwningPointer<DescriptorInterface<DESC>> CreateWithSameTypeAs(
+      const DescriptorInterface<DESC> &source, Rank rank, Attribute attribute) {
+    return OwningPointer<DescriptorInterface<DESC>>{
+        static_cast<DescriptorInterface<DESC> *>(&(*DESC::CreateWithSameTypeAs(
+            static_cast<const DESC &>(source), rank, attribute)))};
+  }
+
+  // Allocate a descriptor for which the type is fully defined:
+  //  - set bounds information (the lower bounds are set to to 1)
+  //  - get storage for the entity described by the descriptor and
+  //    make it available to the descriptor.
+  //  The data is not initialized (even for derived type).
+  void Allocate(const SubscriptArray &extents) { DESC::Allocate(extents); }
```

Comment:
Will the results of transformational intrinsics require dynamic allocation for their results?

Reply:
I did not consider this here, but I think they should not so that we can avoid temporaries as much as possible.
I will try to split the logical and memory part of reshape to not have to allocate every results.
Then it would make sens to implement the transformational as member function of descriptor (that operates directly on their results). Do you agree ?

Reply:
We can avoid temporaries in the runtime for many intrinsics using a more complicated API.  Most of the intrinsic functions and intrinsic operations can be viewed as functions of indices into their results.